# MAPseq pipeline Quick Start Guide
We recommend this tutorial for users who have a basic understanding of virtual environments, the command line, and single cell sequencing.

## Pre-requisites
- **Data**: The MAPseq technique combines single cell transcriptome sequencing + feature barcoding (and optionally paired immune repertoire profiling) with chromatin accessibility sequencing + bridged hashtagging using the [ASAPseq protocol](https://cite-seq.com/asapseq/). The inputs to this pipeline are either the **demultiplexed gzipped fastq files** or the **non-demultiplexed bcl files from the flowcell**.
- **Compute Resources**: The mapping and analysis of single cell data can be computationally intensive. We recommend using a machine with **at least 8 CPUs and 32GB of RAM** to run the MAPseq pipeline. Additionally, it currently **only supports Linux machines**.
- **Input configuration and reference files**: There are **files that must be modified before each run** to ensure the proper running of the pipeline. These are discussed in further detail below.

## Setting up the Environment
- step by step instructions on how to set up the working environment

## Running the Pipeline

### 1. Executing a single run
- how to execute the pipeline
- code the user can copy and paste

### 2. Aggregating multiple runs
- how to execute the pipeline
- code the user can copy and paste

## Output and Results
- what kind of output the pipeline generates? (files, visualizations, etc)
- where will the user find the output?
- basic understanding of what the results mean

## Additional Tips and FAQs
- troubleshooting tips for common errors
- how to submit and issue or question